File: CMPS3500_MiniScheme_Starter_Package/procedural/parser.cpp
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga  /
  / ASGT: CHECKPOINT 1                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: parser.cpp                                    /
  / DATE: 03/29/2026                                    /
*/

#include "parser.h"
#include <iostream>

// Handles a single expression
std::vector<std::string> parseTokens(const std::vector<std::string> &input_tokens)
{
    std::vector<std::string> parsed_tokens;

    // Basic check
    if (input_tokens.size() == 0)
    {
        return parsed_tokens;
    }

    // Check for parentheses
    // if it starts with '(' it should end with ')' 
    if (input_tokens[0] == "(")
    {

        if (input_tokens[input_tokens.size() - 1] != ")")
        {
            std::cerr << "Parse error: missing closing parenthesis\n";
            return parsed_tokens;
        }

        // Extract everything inside parentheses
        for (int i = 1; i < (int)input_tokens.size() - 1; i++)
        {
            parsed_tokens.push_back(input_tokens[i]);
        }
    }
    else
    {
        // Single token (like x)
        parsed_tokens.push_back(input_tokens[0]);
    }

    return parsed_tokens;
}

// Splits multiple expressions from token list
std::vector<std::vector<std::string>> splitExpressions(const std::vector<std::string> &tokens)
{
    std::vector<std::vector<std::string>> expressions;
    std::vector<std::string> current;

    int depth = 0;
    int i;

    for (i = 0; i < (int)tokens.size(); i++)
    {

        if (tokens[i] == "(")
        {
            if (depth == 0)
            {
                current.clear();
            }
            depth++;
        }

        if (depth > 0)
        {
            current.push_back(tokens[i]);
        }

        if (tokens[i] == ")")
        {
            depth--;

            if (depth == 0)
            {
                expressions.push_back(current);
            }
        }

        if (depth == 0 && tokens[i] != "(" && tokens[i] != ")")
        {
            std::vector<std::string> single;
            single.push_back(tokens[i]);
            expressions.push_back(single);
        }
    }

    return expressions;
}

File: CMPS3500_MiniScheme_Starter_Package/oop/Parser.java
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga /
  / ASGT: CHECKPOINT 2                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: Parser.java                                   /
  / DATE: 04/11/2026                                    /
*/

import java.util.ArrayList;
import java.util.List;

public class Parser {
    public static List<String> parseTokens(List<String> inputTokens){
        List<String> parsedTokens = new ArrayList<>();

        if(inputTokens.isEmpty()){
            return parsedTokens;
        }

        if(inputTokens.get(0).equals("(")){
            if(inputTokens.get(inputTokens.size() - 1).equals(")")){
                System.err.println("Parse error: missing closing parenthesis");
                return parsedTokens;
            }
            for(int i = 1; i < inputTokens.size() - 1; i++){
                parsedTokens.add(inputTokens.get(i));
            }
        }
        else{
            parsedTokens.add(inputTokens.get(0));
        }

        return parsedTokens;
    }

    public static List<List<String>> splitExpressions(List<String> tokens){
        List<List<String>> expressions = new ArrayList<>();
        List<String> current = new ArrayList<>();
        int depth = 0;
        int i;

        for(i = 0; i < tokens.size(); i++){
            if(tokens.get(i).equals("(")){
                if(depth == 0){
                    current.clear();
                }
                depth++;
            }

            if(depth > 0){
                current.add(tokens.get(i));
            }

            if(tokens.get(i).equals(")")){
                depth--;

                if(depth == 0){
                    expressions.add(current);
                }
            }

            if(depth == 0 && !tokens.get(i).equals("(") && !tokens.get(i).equals(")")){
                List<String> single = new ArrayList<>();
                single.add(tokens.get(i));
                expressions.add(single);
            }
        }
        
        return expressions;
    }
}